provider-example/src/kx.rs

```diff
@@ -1,39 +1,24 @@
-use crypto::SupportedGroup;
+use crypto::SupportedKxGroup;
 use rustls::crypto;
 
 pub struct KeyExchange {
     priv_key: x25519_dalek::EphemeralSecret,
     pub_key: x25519_dalek::PublicKey,
 }
 
-impl crypto::KeyExchange for KeyExchange {
-    type SupportedGroup = X25519;
-
-    fn start(
-        name: rustls::NamedGroup,
-        _: &[&'static Self::SupportedGroup],
-    ) -> Result<Self, crypto::KeyExchangeError> {
-        if name == rustls::NamedGroup::X25519 {
-            let priv_key = x25519_dalek::EphemeralSecret::random_from_rng(rand_core::OsRng);
-            let pub_key = (&priv_key).into();
-            Ok(KeyExchange { priv_key, pub_key })
-        } else {
-            Err(crypto::KeyExchangeError::UnsupportedGroup)
-        }
-    }
-
-    fn complete<T>(
-        self,
+impl crypto::ActiveKeyExchange for KeyExchange {
+    fn complete(
+        self: Box<KeyExchange>,
         peer: &[u8],
-        f: impl FnOnce(&[u8]) -> Result<T, ()>,
-    ) -> Result<T, rustls::Error> {
+        sink: &mut dyn FnMut(&[u8]),
```

Comment:
Why not `FnOnce`? The current version of the call sites looks pretty bad with all the unwrapping, I think's probably a way to make this cleaner?

Reply:
`FnOnce` consumes `self`, which means the caller of it needs the original closure rather than a reference to it. But the caller is the implementer of a trait, so can't be generic over the several possible `FnOnce` closures. The other alternative would be to have a `Box<dyn FnOnce>` which can be consumed, or have `complete` return the secret value (which would solve all the other shenanigans, but has other costs).

Reply:
Can you make a quick list of options you see and their pros/cons?

Reply:
- result passed into `&mut dyn FnMut(&[u8])` (currently)
  - 🟢 pro: no extra allocations
  - 🔴 con: lots of shenanigans needed to integrate with TLS1.3 key schedule session types; this will panic if _actually_ called twice
- result passed into a `Box<FnOnce(&[u8])>`
  - 🟢 pro: `FnOnce` is the correct type for the meaning we want
  - 🔴 con: extra allocation
- result directly returned from `finish()`
  - 🟢 pro: way simpler, especially for the TLS1.3 key schedule session types.
  - 🔴 con: extra copy
  - 🔴 con: extra allocation
  - 🔴 con: doing an end run around carefully designed underlying type from crypto libaries (eg, x25519-dalek has https://docs.rs/x25519-dalek/latest/x25519_dalek/struct.SharedSecret.html which impls Zeroize, ring passes the secret into a closure)

Reply:
> result passed into a `Box<FnOnce(&[u8])>`

having tried this i'm no longer sure it will work, because the closure has its lifetime (and of its captures)  extended by the box. that means it's no longer possible to use anything captured after the `complete` call.

going to go for the simple option

Reply:
> going to go for the simple option

added this as a seperate commit for incremental review; then i will squash it into the first commit